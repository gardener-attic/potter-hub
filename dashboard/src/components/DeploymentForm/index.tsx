import DeploymentForm from "./DeploymentForm";

export default DeploymentForm;
