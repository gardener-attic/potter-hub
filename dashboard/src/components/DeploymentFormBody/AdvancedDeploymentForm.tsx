import * as React from "react";
import AceEditor from "react-ace";

import "ace-builds/src-noconflict/mode-yaml";
import "ace-builds/src-noconflict/theme-xcode";

export interface IAdvancedDeploymentForm {
  appValues?: string;
  handleValuesChange: (value: string) => void;
}

class AdvancedDeploymentForm extends React.Component<IAdvancedDeploymentForm> {
  public render() {
    return (
      <div className="margin-t-normal">
        <AceEditor
          mode="yaml"
          theme="xcode"
          width="100%"
          onChange={this.props.handleValuesChange}
          setOptions={{ showPrintMargin: false }}
          editorProps={{ $blockScrolling: Infinity }}
          value={this.props.appValues}
          className="editor"
        />
        <p>
          <b>Note:</b> Only comments from the original chart values will be preserved.
        </p>
      </div>
    );
  }
}

export default AdvancedDeploymentForm;
