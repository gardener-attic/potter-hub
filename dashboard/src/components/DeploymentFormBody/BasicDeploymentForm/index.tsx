import BasicDeploymentForm from "./BasicDeploymentForm";

export default BasicDeploymentForm;
