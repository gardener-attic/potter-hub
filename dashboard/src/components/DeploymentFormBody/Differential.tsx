import * as jsdiff from "diff";
import { Diff2Html } from "diff2html";
import * as React from "react";

import "diff2html/dist/diff2html.css";

export interface IDifferentialProps {
  title: string;
  oldValues: string;
  newValues: string;
  emptyDiffText: string;
}

class Differential extends React.Component<IDifferentialProps> {
  public render = () => {
    const { oldValues, newValues, title, emptyDiffText } = this.props;
    const sdiff = jsdiff.createPatch(title, oldValues, newValues);
    const outputHtml = Diff2Html.getPrettyHtml(sdiff, {
      inputFormat: "diff",
      showFiles: false,
      matching: "lines",
      maxLineSizeInBlockForComparison: 20,
    });
    return (
      <div className="diff">
        {oldValues === newValues ? (
          <p>{emptyDiffText}</p>
        ) : (
          <div dangerouslySetInnerHTML={{ __html: outputHtml }} />
        )}
      </div>
    );
  };
}

export default Differential;
