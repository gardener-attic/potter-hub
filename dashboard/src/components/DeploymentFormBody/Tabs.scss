.react-tabs {
  &__tab {
    color: black;
    &--selected {
      color: #1598cb;
    }
  }
}

.editor {
  z-index: 0;
}

.diff {
  height: 500px;
  overflow: auto;
}

.diff pre {
  background-color: inherit;
  padding: 0;
  font: 12px / normal "Monaco", "Menlo", "Ubuntu Mono", "Consolas", "source-code-pro", monospace;
  line-height: 12px;
}

.diff table td {
  padding: 0;
}

.diff table td[class*="linenumber"] {
  position: inherit;
}

.diff table td div[class*="code-line"] {
  padding: 0 0.3em;
  min-width: 58em;
}
