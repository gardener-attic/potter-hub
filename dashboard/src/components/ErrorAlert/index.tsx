import ErrorAlertHeader from "./ErrorAlertHeader";
import ErrorSelector from "./ErrorSelector";
import MessageAlert from "./MessageAlert";
import NotFoundErrorAlert from "./NotFoundErrorAlert";
import PermissionsErrorAlert from "./PermissionsErrorAlert";
import ServiceBrokersNotFoundAlert from "./ServiceBrokersNotFoundAlert";
import ServiceCatalogNotInstalledAlert from "./ServiceCatalogNotInstalledAlert";
import UnexpectedErrorAlert from "./UnexpectedErrorAlert";

export {
  ErrorAlertHeader,
  ErrorSelector,
  MessageAlert,
  NotFoundErrorAlert,
  PermissionsErrorAlert,
  ServiceBrokersNotFoundAlert,
  ServiceCatalogNotInstalledAlert,
  UnexpectedErrorAlert,
};
