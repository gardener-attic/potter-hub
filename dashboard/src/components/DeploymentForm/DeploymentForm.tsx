import { RouterAction } from "connected-react-router";
import * as React from "react";

import { JSONSchema4 } from "json-schema";
import { IAppRepository, IChartState, IChartVersion } from "../../shared/types";
import { getInfoAnnotation } from "../../shared/utils";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import InfoBox from "../InfoBox";
import LoadingWrapper from "../LoadingWrapper";

import "react-tabs/style/react-tabs.css";

export interface IDeploymentFormProps {
  kubeappsNamespace: string;
  chartID: string;
  chartVersion: string;
  error: Error | undefined;
  selected: IChartState["selected"];
  deployChart: (
    version: IChartVersion,
    releaseName: string,
    namespace: string,
    values?: string,
    schema?: JSONSchema4,
  ) => Promise<boolean>;
  push: (location: string) => RouterAction;
  fetchChartVersions: (id: string) => void;
  getChartVersion: (id: string, chartVersion: string) => void;
  namespace: string;
  repos: IAppRepository[];
  fetchRepos: () => void;
  repoErrors: {
    create?: Error;
    delete?: Error;
    fetch?: Error;
    update?: Error;
  };
  generateReleaseName: () => string
}

export interface IDeploymentFormState {
  isDeploying: boolean;
  releaseName: string;
  // Name of the release that was submitted for creation
  // This is different than releaseName since it is also used in the error banner
  // and we do not want to use releaseName since it is controller by the form field.
  latestSubmittedReleaseName: string;
  appValues: string;
  valuesModified: boolean;
}

class DeploymentForm extends React.Component<IDeploymentFormProps, IDeploymentFormState> {
  public state: IDeploymentFormState = {
    releaseName: this.props.generateReleaseName(),
    appValues: this.props.selected.values || "",
    isDeploying: false,
    latestSubmittedReleaseName: "",
    valuesModified: false,
  };

  public componentDidMount() {
    this.props.fetchChartVersions(this.props.chartID);
    this.props.fetchRepos();
  }

  public componentDidUpdate(prevProps: IDeploymentFormProps) {
    if (prevProps.selected.version !== this.props.selected.version && !this.state.valuesModified) {
      this.setState({ appValues: this.props.selected.values || "" });
    }
  }

  public render() {
    const { namespace } = this.props;
    if (this.props.error) {
      return (
        <ErrorSelector
          error={this.props.error}
          namespace={namespace}
          action="create"
          resource={this.state.latestSubmittedReleaseName}
        />
      );
    }
    if (this.state.isDeploying) {
      return <LoadingWrapper />;
    }

    const repoName = this.props.chartID.split("/")[0];
    let repo = {} as IAppRepository;
    for (const r of this.props.repos) {
      if (r.metadata.name === repoName) {
        repo = r;
        break;
      }
    }

    return (
      <form className="container padding-b-bigger" onSubmit={this.handleDeploy}>
        <div className="row">
          {this.props.repoErrors.fetch && (
            <div className="col-8">
              <ErrorSelector
                error={this.props.repoErrors.fetch}
                action="get"
                resource="appRepositories"
              />
            </div>
          )}
          {!this.props.repoErrors.fetch && (
            <div className="col-8">
              <InfoBox msg={getInfoAnnotation(repo)} />
            </div>
          )}
          <div className="col-12">
            <h2>{this.props.chartID}</h2>
          </div>
          <div className="col-8">
            <div>
              <label htmlFor="releaseName">Name</label>
              <input
                id="releaseName"
                pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
                title="Use lower case alphanumeric characters, '-' or '.'"
                onChange={this.handleReleaseNameChange}
                value={this.state.releaseName}
                required={true}
              />
            </div>
            <DeploymentFormBody
              chartID={this.props.chartID}
              chartVersion={this.props.chartVersion}
              namespace={this.props.namespace}
              selected={this.props.selected}
              push={this.props.push}
              getChartVersion={this.props.getChartVersion}
              setValues={this.handleValuesChange}
              appValues={this.state.appValues}
              setValuesModified={this.setValuesModified}
            />
          </div>
        </div>
      </form>
    );
  }

  public handleValuesChange = (value: string) => {
    this.setState({ appValues: value });
  };

  public setValuesModified = () => {
    this.setState({ valuesModified: true });
  };

  public handleDeploy = async (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();
    const { selected, deployChart, push, namespace } = this.props;
    const { releaseName, appValues } = this.state;

    this.setState({ isDeploying: true, latestSubmittedReleaseName: releaseName });
    if (selected.version) {
      const deployed = await deployChart(
        selected.version,
        releaseName,
        namespace,
        appValues,
        selected.schema,
      );
      this.setState({ isDeploying: false });
      if (deployed) {
        push(`/apps/ns/${namespace}/${releaseName}`);
      }
    }
  };

  public handleReleaseNameChange = (e: React.FormEvent<HTMLInputElement>) => {
    this.setState({ releaseName: e.currentTarget.value });
  };
}

export default DeploymentForm;
