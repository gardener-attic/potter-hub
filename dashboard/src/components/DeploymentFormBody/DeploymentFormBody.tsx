import { RouterAction } from "connected-react-router";
import * as jsonpatch from "fast-json-patch";
import * as React from "react";
import { AlertOctagon } from "react-feather";
import { Tab, TabList, TabPanel, Tabs } from "react-tabs";

import { definedNamespaces } from "../../shared/Namespace";
import { retrieveBasicFormParams, setValue } from "../../shared/schema";
import { IBasicFormParam, IChartState } from "../../shared/types";
import { getValueFromEvent } from "../../shared/utils";
import ConfirmDialog from "../ConfirmDialog";
import { ErrorSelector } from "../ErrorAlert";
import Hint from "../Hint";
import LoadingWrapper from "../LoadingWrapper";
import AdvancedDeploymentForm from "./AdvancedDeploymentForm";
import BasicDeploymentForm from "./BasicDeploymentForm";

import "react-tabs/style/react-tabs.css";
import Differential from "./Differential";
import "./Tabs.css";

export interface IDeploymentFormBodyProps {
  chartID: string;
  chartVersion: string;
  deployedValues?: string;
  namespace: string;
  releaseVersion?: string;
  selected: IChartState["selected"];
  appValues: string;
  push: (location: string) => RouterAction;
  goBack?: () => RouterAction;
  getChartVersion: (id: string, chartVersion: string) => void;
  setValues: (values: string) => void;
  setValuesModified: () => void;
}

export interface IDeploymentFormBodyState {
  basicFormParameters: IBasicFormParam[];
  restoreDefaultValuesModalIsOpen: boolean;
  modifications?: jsonpatch.Operation[];
}

class DeploymentFormBody extends React.Component<
  IDeploymentFormBodyProps,
  IDeploymentFormBodyState
> {
  public state: IDeploymentFormBodyState = {
    basicFormParameters: [],
    restoreDefaultValuesModalIsOpen: false,
  };

  public componentDidMount() {
    const { chartID, getChartVersion, chartVersion } = this.props;
    getChartVersion(chartID, chartVersion);
  }

  public componentDidUpdate = (prevProps: IDeploymentFormBodyProps) => {
    const { chartID, chartVersion, getChartVersion, selected, appValues } = this.props;

    if (chartVersion !== prevProps.chartVersion) {
      // New version detected
      getChartVersion(chartID, chartVersion);
      return;
    }

    if (
      // If the selected schema changes
      prevProps.selected.schema !== selected.schema ||
      // or the selected values (because it's a different version)
      prevProps.selected.values !== selected.values ||
      // or the current values (and we hadn't process those values yet)
      (prevProps.appValues !== appValues && prevProps.appValues.length === 0)
    ) {
      // Parse the basic parameters
      this.setState({
        basicFormParameters: retrieveBasicFormParams(appValues, selected.schema),
      });
    }
  };

  public render() {
    const { selected, chartID, chartVersion, goBack, namespace } = this.props;
    const { version, versions } = selected;
    if (selected.error) {
      return (
        <ErrorSelector error={selected.error} resource={`Chart "${chartID}" (${chartVersion})`} />
      );
    }
    if (!version || !versions.length) {
      return <LoadingWrapper />;
    }
    return (
      <div>
        <ConfirmDialog
          modalIsOpen={this.state.restoreDefaultValuesModalIsOpen}
          loading={false}
          confirmationText={"Are you sure you want to restore the default chart values?"}
          confirmationButtonText={"Restore"}
          onConfirm={this.restoreDefaultValues}
          closeModal={this.closeRestoreDefaultValuesModal}
        />
        <div>
          <label htmlFor="chartVersion">Version</label>
          <select
            id="chartVersion"
            onChange={this.handleChartVersionChange}
            value={version.attributes.version}
            required={true}
          >
            {versions.map(v => (
              <option key={v.id} value={v.attributes.version}>
                {`Chart: ${v.attributes.version} | App: ${v.attributes.app_version || "-"} `} 
                {this.props.releaseVersion && v.attributes.version === this.props.releaseVersion
                  ? "(current)"
                  : ""}
              </option>
            ))}
          </select>
        </div>
        {this.renderTabs()}
        <div className="margin-t-big">
          <button
            className="button button-primary"
            type="submit"
            disabled={!namespace || namespace === definedNamespaces.all}
          >
            Submit
          </button>
          <button className="button" type="button" onClick={this.openRestoreDefaultValuesModal}>
            Restore Chart Defaults
          </button>
          {goBack && (
            <button className="button" type="button" onClick={goBack}>
              Back
            </button>
          )}
          {(!namespace || namespace === definedNamespaces.all) && (
            <div className="container-tiny" style={{ display: "flex", marginTop: "10px" }}>
              <div className="alert alert-warning AlertBox" role="alert">
                <div className="AlertBoxIconContainer">
                  <AlertOctagon />
                </div>
                Please select a namespace using the namespace selector from the top menu.
              </div>
            </div>
          )}
        </div>
      </div>
    );
  }

  private handleChartVersionChange = (e: React.FormEvent<HTMLSelectElement>) => {
    // TODO(andres): This requires refactoring. Currently, the deploy and upgrade
    // forms behave differently. In the deployment form, a change in the version
    // changes the route but in the case of the upgrade it only changes the state
    const isUpgradeForm = !!this.props.releaseVersion;

    if (isUpgradeForm) {
      const { chartID, getChartVersion } = this.props;
      getChartVersion(chartID, e.currentTarget.value);
    } else {
      this.props.push(
        `/apps/ns/${this.props.namespace}/new/${this.props.chartID}/versions/${e.currentTarget.value}`,
      );
    }
  };

  private handleValuesChange = (value: string) => {
    this.props.setValues(value);
    this.props.setValuesModified();
  };

  private refreshBasicParameters = () => {
    this.setState({
      basicFormParameters: retrieveBasicFormParams(
        this.props.appValues,
        this.props.selected.schema,
      ),
    });
  };

  private renderTabs = () => {
    return (
      <div className="margin-t-normal">
        <Tabs>
          <TabList>
            {this.shouldRenderBasicForm() && (
              <Tab onClick={this.refreshBasicParameters}>
                Form{" "}
                <Hint reactTooltipOpts={{ delayHide: 100 }} id="basicFormHelp">
                  <span>
                    This form has been automatically generated based on the chart schema.
                    <br />
                    This feature is currently in a beta state. If you find an issue please report it{" "}
                    <a
                      target="_blank"
                      rel="noopener noreferrer"
                      href="https://github.com/gardener/potter-hub/issues/new/choose"
                    >
                      here.
                    </a>
                  </span>
                </Hint>
              </Tab>
            )}
            <Tab>Values (YAML)</Tab>
            <Tab>Changes</Tab>
          </TabList>
          {this.shouldRenderBasicForm() && (
            <TabPanel>
              <BasicDeploymentForm
                params={this.state.basicFormParameters}
                handleBasicFormParamChange={this.handleBasicFormParamChange}
                appValues={this.props.appValues}
                handleValuesChange={this.handleValuesChange}
              />
            </TabPanel>
          )}
          <TabPanel>
            <AdvancedDeploymentForm
              appValues={this.props.appValues}
              handleValuesChange={this.handleValuesChange}
            />
          </TabPanel>
          <TabPanel>{this.renderDiff()}</TabPanel>
        </Tabs>
      </div>
    );
  };

  private handleBasicFormParamChange = (param: IBasicFormParam) => {
    return (e: React.FormEvent<any>) => {
      this.props.setValuesModified();
      const value = getValueFromEvent(e);
      this.setState({
        // Change param definition
        basicFormParameters: this.state.basicFormParameters.map(p =>
          p.path === param.path ? { ...param, value } : p,
        ),
      });
      // Change raw values
      this.props.setValues(setValue(this.props.appValues, param.path, value));
    };
  };

  // The basic form should be rendered if there are params to show
  private shouldRenderBasicForm = () => {
    return Object.keys(this.state.basicFormParameters).length > 0;
  };

  private closeRestoreDefaultValuesModal = () => {
    this.setState({ restoreDefaultValuesModalIsOpen: false });
  };

  private openRestoreDefaultValuesModal = () => {
    this.setState({ restoreDefaultValuesModalIsOpen: true });
  };

  private restoreDefaultValues = () => {
    if (this.props.selected.values) {
      this.props.setValues(this.props.selected.values);
      this.setState({
        basicFormParameters: retrieveBasicFormParams(
          this.props.selected.values,
          this.props.selected.schema,
        ),
      });
    }
    this.setState({ restoreDefaultValuesModalIsOpen: false });
  };

  private renderDiff = () => {
    let oldValues = "";
    let title = "";
    let emptyDiffText = "";
    if (this.props.deployedValues) {
      // If there are already some deployed values (upgrade scenario)
      // We compare the values from the old release and the new one
      oldValues = this.props.deployedValues;
      title = "Difference from deployed version";
      emptyDiffText = "The values for the new release are identical to the deployed version.";
    } else {
      // If it's a new deployment, we show the different from the default
      // values for the selected version
      oldValues = this.props.selected.values || "";
      title = "Difference from chart defaults";
      emptyDiffText = "No changes detected from chart defaults.";
    }
    return (
      <Differential
        title={title}
        oldValues={oldValues}
        newValues={this.props.appValues}
        emptyDiffText={emptyDiffText}
      />
    );
  };
}

export default DeploymentFormBody;
