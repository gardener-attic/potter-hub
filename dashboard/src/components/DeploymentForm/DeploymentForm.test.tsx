import { mount, shallow } from "enzyme";
import * as React from "react";

import {
  IAppRepository,
  IChartState,
  IChartVersion,
  UnprocessableEntity,
} from "../../shared/types";
import DeploymentFormBody from "../DeploymentFormBody/DeploymentFormBody";
import { ErrorSelector } from "../ErrorAlert";
import DeploymentForm from "./DeploymentForm";

const releaseName = "my-release";
const defaultProps = {
  kubeappsNamespace: "kubeapps",
  chartID: "foo",
  chartVersion: "1.0.0",
  error: undefined,
  selected: {} as IChartState["selected"],
  deployChart: jest.fn(),
  push: jest.fn(),
  fetchChartVersions: jest.fn(),
  getChartVersion: jest.fn(),
  namespace: "default",
  repos: [] as IAppRepository[],
  fetchRepos: () => jest.fn(),
  repoErrors: {
    create: undefined,
    delete: undefined,
    fetch: undefined,
    update: undefined,
  },
  generateReleaseName: jest.fn()
};
const versions = [{ id: "foo", attributes: { version: "1.2.3" } }] as IChartVersion[];

beforeEach(() => {
  defaultProps.generateReleaseName = jest.fn(() => releaseName);
});

it("fetches the available versions", () => {
  const fetchChartVersions = jest.fn();
  shallow(<DeploymentForm {...defaultProps} fetchChartVersions={fetchChartVersions} />);
  expect(fetchChartVersions).toHaveBeenCalledWith(defaultProps.chartID);
});

describe("renders an error", () => {
  it("renders a custom error if the deployment failed", () => {
    const wrapper = shallow(
      <DeploymentForm
        {...defaultProps}
        selected={
          {
            version: { attributes: {} },
            versions: [{ id: "foo", attributes: {} }],
          } as IChartState["selected"]
        }
        error={new UnprocessableEntity("wrong format!")}
      />,
    );
    wrapper.setState({ latestSubmittedReleaseName: "my-app" });
    expect(wrapper.find(ErrorSelector).exists()).toBe(true);
    expect(wrapper.find(ErrorSelector).html()).toContain(
      "Sorry! Something went wrong processing my-app",
    );
    expect(wrapper.find(ErrorSelector).html()).toContain("wrong format!");
  });

  it("the error does not change if the release name changes", () => {
    const expectedErrorMsg = "Sorry! Something went wrong processing my-app";

    const wrapper = shallow(
      <DeploymentForm
        {...defaultProps}
        selected={
          {
            version: { attributes: {} },
            versions: [{ id: "foo", attributes: {} }],
          } as IChartState["selected"]
        }
        error={new UnprocessableEntity("wrong format!")}
      />,
    );

    wrapper.setState({ latestSubmittedReleaseName: "my-app" });
    expect(wrapper.find(ErrorSelector).exists()).toBe(true);
    expect(wrapper.find(ErrorSelector).html()).toContain(expectedErrorMsg);
    wrapper.setState({ releaseName: "another-app" });
    expect(wrapper.find(ErrorSelector).html()).toContain(expectedErrorMsg);
  });
});

it("renders the full DeploymentForm", () => {
  const wrapper = shallow(
    <DeploymentForm {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  expect(wrapper).toMatchSnapshot();
});

it("renders a release name by default, relying in Monickers output", () => {
  defaultProps.generateReleaseName = jest.fn().mockImplementationOnce(() => "foo").mockImplementationOnce(() => "bar")

  let wrapper = shallow(
    <DeploymentForm {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  const name1 = wrapper.state("releaseName") as string;
  expect(name1).toBe("foo");

  // When reloading the name should change
  wrapper = shallow(
    <DeploymentForm {...defaultProps} selected={{ versions, version: versions[0] }} />,
  );
  const name2 = wrapper.state("releaseName") as string;
  expect(name2).toBe("bar");
});

it("forwards the appValues when modified", () => {
  const wrapper = shallow(<DeploymentForm {...defaultProps} />);
  const handleValuesChange: (v: string) => void = wrapper
    .find(DeploymentFormBody)
    .prop("setValues");
  handleValuesChange("foo: bar");

  expect(wrapper.state("appValues")).toBe("foo: bar");
  expect(wrapper.find(DeploymentFormBody).prop("appValues")).toBe("foo: bar");
});

it("forwards the valuesModifed property", () => {
  const wrapper = shallow(<DeploymentForm {...defaultProps} />);
  const handleValuesModified: (v: string) => void = wrapper
    .find(DeploymentFormBody)
    .prop("setValuesModified");
  handleValuesModified("foo: bar");

  expect(wrapper.state("valuesModified")).toBe(true);
});

it("triggers a deployment when submitting the form", done => {
  const namespace = "default";
  const appValues = "foo: bar";
  const schema = { properties: { foo: { type: "string", form: true } } };
  const deployChart = jest.fn(() => true);
  const push = jest.fn();
  const wrapper = mount(
    <DeploymentForm
      {...defaultProps}
      selected={{ versions, version: versions[0], schema }}
      deployChart={deployChart}
      push={push}
      namespace={namespace}
    />,
  );
  wrapper.setState({ appValues });
  wrapper.find("form").simulate("submit");
  expect(deployChart).toHaveBeenCalledWith(versions[0], releaseName, namespace, appValues, schema);
  setTimeout(() => {
    expect(push).toHaveBeenCalledWith("/apps/ns/default/my-release");
    done();
  }, 1);
});
