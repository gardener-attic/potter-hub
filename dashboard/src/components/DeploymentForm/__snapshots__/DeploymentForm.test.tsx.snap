// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the full DeploymentForm 1`] = `
<form
  className="container padding-b-bigger"
  onSubmit={[Function]}
>
  <div
    className="row"
  >
    <div
      className="col-8"
    >
      <InfoBox
        msg=""
      />
    </div>
    <div
      className="col-12"
    >
      <h2>
        foo
      </h2>
    </div>
    <div
      className="col-8"
    >
      <div>
        <label
          htmlFor="releaseName"
        >
          Name
        </label>
        <input
          id="releaseName"
          onChange={[Function]}
          pattern="[a-z0-9]([-a-z0-9]*[a-z0-9])?(\\\\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*"
          required={true}
          title="Use lower case alphanumeric characters, '-' or '.'"
          value="my-release"
        />
      </div>
      <DeploymentFormBody
        appValues=""
        chartID="foo"
        chartVersion="1.0.0"
        getChartVersion={[MockFunction]}
        namespace="default"
        push={[MockFunction]}
        selected={
          Object {
            "version": Object {
              "attributes": Object {
                "version": "1.2.3",
              },
              "id": "foo",
            },
            "versions": Array [
              Object {
                "attributes": Object {
                  "version": "1.2.3",
                },
                "id": "foo",
              },
            ],
          }
        }
        setValues={[Function]}
        setValuesModified={[Function]}
      />
    </div>
  </div>
</form>
`;
